### ISSUE TITLE
The rightmost widget in the top wibar disappears sometimes

### ISSUE THREAD

**Issue by p-himik** [awesomeWM/awesome#3089]
Output of `awesome --version`:
```
awesome v4.3 (Too long)
 • Compiled against Lua 5.3.3 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.6
 • LGI version: 0.9.2
```
**How to reproduce the issue:**
Honestly - no idea. It happens under _very_ particular circumstances that I just randomly happened upon.
Here's what I have right now. The flag is just an image widget, then goes the system tray, and then should go the calendar widget.
![image](https://user-images.githubusercontent.com/4410314/80890740-f0874980-8cc7-11ea-832e-0de830ca6e91.png)
As soon as I remove _any_ icon from the tray or any widget to the left, the calendar comes back:
![image](https://user-images.githubusercontent.com/4410314/80890775-2a585000-8cc8-11ea-8ed2-f94a5869020b.png)

When I restart Awesome, the calendar widget appears for a split second and then disappears again.
That blank area is not interactive at all - seems as if the widget is not even there. It can happen with other widgets as well - in my case, it works with `layoutbox` but I can't make it work with `textbox` (tried texts of different lengths) or with `separator` (tried forcing the width from 1 to 24). Maybe I just didn't stumble upon that magic width.

My current workaround is to just add an empty `textbox` as the very right.

**Comment by actionless** [comment]
could you make a full screenshot of the panel before and after?

i have some theory about that behavior, but need more input

**Comment by p-himik** [comment]
Sure, in this case it's `layoutbox` that's disappeared:
![image](https://user-images.githubusercontent.com/4410314/80891704-68f10900-8cce-11ea-85a5-343e44f74c99.png)
And now it's back:
![image](https://user-images.githubusercontent.com/4410314/80891711-81612380-8cce-11ea-8911-fb9ede683b36.png)

**Comment by actionless** [comment]
my theory was what you might have some fixed-size widget in the center/left side, but it seems to be not an issue

for example, when i have too much widget alerts on the left, they don't have any place to render:

![2020-05-02--1588452301_1920x41_scrot](https://user-images.githubusercontent.com/1655669/80891804-bcf7ef80-8cc6-11ea-9b5f-2c4f4dda3c08.png)


you can see what the "8" digit just disappers on the next screen as it doesn't have free space to render:

![2020-05-02--1588452333_1920x47_scrot](https://user-images.githubusercontent.com/1655669/80891816-cd0fcf00-8cc6-11ea-9b9f-82a9b99dfaba.png)


i still think, what in your situation, playing with flex/fixed/align layouts combinations might solve the problem

**Comment by actionless** [comment]
ah, and also file:///usr/share/doc/awesome/doc/widget_layouts/wibox.layout.fixed.html#fill_space and file:///usr/share/doc/awesome/doc/widget_layouts/wibox.layout.align.html#expand

**Comment by p-himik** [comment]
Thanks for checking. Yeah, it's definitely not the lack of space at play here as adding additional stuff solves the issue.

**Comment by actionless** [comment]
> as adding additional stuff solves the issue.

looks like a situation where playing with expand and/or fill_space might help

**Comment by p-himik** [comment]
As I mentioned - I have already solved the issue on my side by using a dummy widget. And I already have a `tasklist` there that should take up all the space, so I don't know why `fill_space` should be relevant.

The issue is definitely with Awesome since the rightmost widget should have all the space it needs exactly because `tasklist` can be shrinked.

**Comment by actionless** [comment]
i would start from tasklist, i suspect, in the layout it's taking its initial width, and next right part of the layout became bigger it doesn't have any free space anymore

**Comment by p-himik** [comment]
> and next right part of the layout became bigger as it doesn't have any free space anymore
If that's the case, then new system tray icon would create even more issues, and for everyone. But having an additional icon just makes the widget appear again.

**Comment by actionless** [comment]
could you show part of the config, there you adding tasklist to the wibar?

**Comment by p-himik** [comment]
Sure:
```lua
    -- Add widgets to the wibox
    s.mywibox:setup {
        layout = wibox.layout.align.horizontal,
        {
            -- Left widgets
            layout = wibox.layout.fixed.horizontal,
            mylauncher,
            s.mytaglist,
            s.mypromptbox,
        },
        s.mytasklist, -- Middle widget
        {
            -- Right widgets
            layout = wibox.layout.fixed.horizontal,
            spacing = math.ceil(size / 5),
            {
                widget = cpuwidget,
                color = {
                    type = "linear",
                    from = { 0, 0 },
                    to = { 0, 40 },
                    stops = { { 0, "#FF5656" }, { 0.5, "#88A175" }, { 1, "#AECF96" } }
                },
                background_color = "#494B4F",
                forced_width = size * 2;
            },
            wibox.widget {
                {
                    widget = memwidget,
                    color = {
                        type = "linear",
                        from = { 0, 0 },
                        to = { 20, 0 },
                        stops = { { 0, "#AECF96" }, { 0.5, "#88A175" }, { 1, "#FF5656" } }
                    },
                    background_color = "#494B4F"
                },
                border_color = nil,
                forced_width = math.ceil(size / 3),
                direction = 'east',
                layout = wibox.container.rotate
            },
            wibox.widget {
                forced_width = math.ceil(size * 1.5),
                widget = apw.progressbar
            },
            --batwidget,
            vpnwidget,
            kbdwidget,
            wibox.widget.systray(),
            calendarwidget,
            s.mylayoutbox,
            -- TODO: Remove when https://github.com/awesomeWM/awesome/issues/3089 is fixed.
            wibox.widget.textbox(''),
        },
    }
```

**Comment by actionless** [comment]
does adding fill_space to the second `layout = wibox.layout.fixed.horizontal,` changing anything?

**Comment by p-himik** [comment]
Yep, it seems to work. Of course, I can't say whether it's guaranteed to work or just happen to work for this particular layout and the number of tray icons.

**Comment by actionless** [comment]
let's keep it open then until you'll be sure

**Comment by p-himik** [comment]
By the very nature of this issue, I can never be sure.

Do you know why the version without `fill_space` works the way it does?
What can be changed in Awesome to prevent such behavior?

**Comment by actionless** [comment]
i wasn't looking deeply, but it seems what layoutbox getting confused in such situation, mb it doesn't have (proper) :fit() function

**Comment by actionless** [comment]
also imagebox-es without forced_ height or width could behave funky inside the layouts

**Comment by p-himik** [comment]
That's interesting, I'll experiment with it then, thanks.

**Comment by actionless** [comment]
please share if you'll get any interesting findings, as i was noticing such behaviors all the time (at least after new layout/widget system was introduced), would be useful if it would be somehow more formalized, so it could be either described better in the docs or fixed

**Comment by p-himik** [comment]
Didn't have a chance to experiment with it yet, but the `fill_space` workaround suddenly stopped working for no apparent reason. Maybe it didn't work back then as well - it just happened to fix the issue for that particular systray icon set.

**Comment by VehementHam** [comment]
I have this issue too
